nvflare/app_opt/pt/quantization/ada_quant.py

```diff
@@ -0,0 +1,115 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import bz2
+import math
+from typing import Any, Optional, Union
+
+import numpy as np
+import torch
+
+
+class AdaQuantizer:
+    def __init__(self, weight: float = 0.01, compression: bool = True) -> None:
+        """Implements the ADAQUANT quantization scheme,
+            for further details refer to the paper https://arxiv.org/abs/2208.05174
+
+        Args:
+            weight: a hyperparameter for the trade-off between quantization size and error
+            compression: whether to compress the resulting integer quantized tensor
+
+        """
+
+        self.weight = weight
+        self.compression = compression
+
+    def quantize(self, values_tensor: torch.Tensor) -> tuple[Union[torch.Tensor, np.ndarray], dict]:
+        values_tensor = values_tensor.cpu()
+        old_values_tensor = values_tensor
+
+        old_tensor_shape = list(old_values_tensor.shape)
+        values_tensor = values_tensor.to(dtype=torch.float64).view(-1)
+        offset = self.get_offset(values_tensor)
+        values_tensor = values_tensor + offset
+        norm = values_tensor.abs().max()
+
+        quant_state = {"offset": offset}
+        res = self.get_number_of_quantization_levels(
+            element_size=old_values_tensor.element_size(), values_tensor=values_tensor
+        )
+        if res is None:
+            return old_values_tensor, {}
+        norm, quantization_level, new_dtype = res
+        if norm == 0.0:
+            return torch.tensor([0], dtype=torch.bool), quant_state | {
+                "tensor_shape": old_tensor_shape,
+            }
+
+        normalized_abs_tensor = values_tensor.abs() / norm
+        quantized_tensor = (normalized_abs_tensor * quantization_level).round().clamp(0, quantization_level)
+        quantized_tensor = quantized_tensor.numpy().astype(dtype=new_dtype)
+        if self.compression:
+            raw_bytes = quantized_tensor.tobytes()
+            compressed_bytes = bz2.compress(raw_bytes)
```

Comment:
any particular reason of using bz2? do we have alternatives if we cannot use it eventually (just in case)?

Reply:
bz2 have a balance between compression rate and speed. We could try zip and others.